PANDAS PROJECT - 2
Lego Sets Analysis

In [1]:
# importing libraries
import pandas as pd

In [2]:
df = pd.read_csv(r"C:\Users\manoj.c\Documents\PANDAS_PROJECT\Pandas-Project-2\Datasets\lego_sets.csv")
df.head()

set_num                        name  year  num_parts    theme_name  \
0    00-1             Weetabix Castle  1970      471.0        Castle   
1  0011-2           Town Mini-Figures  1978        NaN  Supplemental   
2  0011-3  Castle 2 for 1 Bonus Offer  1987        NaN  Lion Knights   
3  0012-1          Space Mini-Figures  1979       12.0  Supplemental   
4  0013-1          Space Mini-Figures  1979       12.0  Supplemental   

  parent_theme  
0     Legoland  
1         Town  
2       Castle  
3        Space  
4        Space

In [3]:
theme = pd.read_csv(r"C:\Users\manoj.c\Documents\PANDAS_PROJECT\Pandas-Project-2\Datasets\parent_themes.csv")
theme.head(10)

id               name  is_licensed
0    1            Technic        False
1   22            Creator        False
2   50               Town        False
3  112             Racers        False
4  126              Space        False
5  147            Pirates        False
6  155  Modular Buildings        False
7  158          Star Wars         True
8  186             Castle        False
9  204      Designer Sets        False

In [ ]:
# merging data using 1st table columns with 2nd table column names.
merged = df.merge(theme, left_on='parent_theme', right_on='name')
merged = merged.drop(columns= 'name_y')
merged.head()

set_num                      name_x  year  num_parts    theme_name  \
0    00-1             Weetabix Castle  1970      471.0        Castle   
1  0011-2           Town Mini-Figures  1978        NaN  Supplemental   
2  0011-3  Castle 2 for 1 Bonus Offer  1987        NaN  Lion Knights   
3  0012-1          Space Mini-Figures  1979       12.0  Supplemental   
4  0013-1          Space Mini-Figures  1979       12.0  Supplemental   

  parent_theme   id  is_licensed  
0     Legoland  411        False  
1         Town   50        False  
2       Castle  186        False  
3        Space  126        False  
4        Space  126        False

In [ ]:

# Taking data only where is_licensed is True
licensed = merged[merged["is_licensed"]]
licensed.head()

set_num                         name_x  year  num_parts  \
44  10018-1                     Darth Maul  2001     1868.0   
45  10019-1    Rebel Blockade Runner - UCS  2001        NaN   
54  10026-1        Naboo Starfighter - UCS  2002        NaN   
57  10030-1  Imperial Star Destroyer - UCS  2002     3115.0   
95  10075-1         Spider-Man Action Pack  2002       25.0   

                 theme_name  parent_theme   id  is_licensed  
44                Star Wars     Star Wars  158         True  
45  Star Wars Episode 4/5/6     Star Wars  158         True  
54      Star Wars Episode 1     Star Wars  158         True  
57  Star Wars Episode 4/5/6     Star Wars  158         True  
95               Spider-Man  Super Heroes  482         True

In [9]:
# gettinng data of only Star_wars
star_wars = licensed[licensed['parent_theme'] == "Star Wars"]
star_wars.head(10)

set_num                           name_x  year  num_parts  \
44   10018-1                       Darth Maul  2001     1868.0   
45   10019-1      Rebel Blockade Runner - UCS  2001        NaN   
54   10026-1          Naboo Starfighter - UCS  2002        NaN   
57   10030-1    Imperial Star Destroyer - UCS  2002     3115.0   
116  10123-1                       Cloud City  2003      707.0   
120  10129-1          Rebel Snowspeeder - UCS  2003     1456.0   
122  10131-1           TIE Fighter Collection  2004        NaN   
125  10134-1  Y-wing Attack Starfighter - UCS  2004        NaN   
126  10143-1                    Death Star II  2005        NaN   
127  10144-1                      Sandcrawler  2005     1679.0   

                  theme_name parent_theme   id  is_licensed  
44                 Star Wars    Star Wars  158         True  
45   Star Wars Episode 4/5/6    Star Wars  158         True  
54       Star Wars Episode 1    Star Wars  158         True  
57   Star Wars Episode 4/5/6    Star Wars  158         True  
116  Star Wars Episode 4/5/6    Star Wars  158         True  
120  Star Wars Episode 4/5/6    Star Wars  158         True  
122  Star Wars Episode 4/5/6    Star Wars  158         True  
125  Star Wars Episode 4/5/6    Star Wars  158         True  
126  Star Wars Episode 4/5/6    Star Wars  158         True  
127  Star Wars Episode 4/5/6    Star Wars  158         True

In [ ]:
# Shapes will get me number of rows & columns
star_wars.shape

(609, 8)

Question 01: What percentage of all licensed sets ever released were Star Wars themed?
Save as a varibale 'the_force' in the form of an integer(eg. 25)

In [14]:
the_force = int((star_wars.shape[0] / licensed.shape[0])*100)
print(the_force)

45


In [16]:
# Data cleaning needs to be done, so lets check is there null values in set_num columns
merged[merged['set_num'].isnull()].shape

(153, 8)

In [22]:
# so lets drop 
licensed = licensed.dropna(subset=['set_num'])
the_force = float((star_wars.shape[0] / licensed.shape[0])*100)
print(the_force)

51.653944020356235


ANSWER : 51.65% 

Question 2: In which year was Star Wars not the most popular licensed theme(in terms of number of sets released that year) ?
save your answer as a variable 'new_era' in the form of an integer (eg.2012)

In [30]:
licensed_sorted = licensed.sort_values('year')
licensed_sorted['count'] = 1
summed_df = licensed_sorted.groupby(['year','parent_theme']).sum().reset_index()
summed_df.head()


year           parent_theme  \
0  1999              Star Wars   
1  2000  Disney's Mickey Mouse   
2  2000              Star Wars   
3  2001           Harry Potter   
4  2001              Star Wars   

                                             set_num  \
0  7140-17141-17150-17151-17110-17101-17121-17130...   
1                     4178-14166-14165-14167-14164-1   
2  8000-18002-17115-13342-13341-17144-17124-17134...   
3  4708-14702-14704-14707-14721-14709-14722-14706...   
4  8008-17146-17126-17127-17106-165034-165030-165...   

                                              name_x  num_parts  \
0  X-wing FighterNaboo FighterTIE Fighter & Y-win...     1384.0   
1  Mickey's Fishing AdventureMickey's Car GarageM...      405.0   
2  Pit Droid™ / Star Wars Pit DroidDestroyer Droi...     2580.0   
3  Hogwarts ExpressThe Final ChallengeThe Chamber...     1284.0   
4  Stormtrooper™TIE FighterBattle Droid CarrierIm...     2949.0   

                                          theme_name    id  is_licensed  count  
0  Star Wars Episode 4/5/6Star Wars Episode 1Star...  2054           13     13  
1  Disney's Mickey MouseDisney's Mickey MouseDisn...  1940            5      5  
2  Star WarsStar WarsStar Wars Episode 1Star Wars...  4108           26     26  
3  Harry PotterSorcerer's StoneSorcerer's StoneSo...  2706           11     11  
4  Star WarsStar Wars Episode 4/5/6Star Wars Epis...  2212           14     14

In [36]:
max_df = summed_df.sort_values('count', ascending=False).drop_duplicates('year')
max_df.sort_values('year',inplace=True)
max_df.head(100)

year  parent_theme                                            set_num  \
0   1999     Star Wars  7140-17141-17150-17151-17110-17101-17121-17130...   
2   2000     Star Wars  8000-18002-17115-13342-13341-17144-17124-17134...   
4   2001     Star Wars  8008-17146-17126-17127-17106-165034-165030-165...   
6   2002     Star Wars  4195641-165153-165106-17103-17203-17204-17133-...   
9   2003     Star Wars  4501-14482-14481-14480-14500-1K4479-1K4487-144...   
12  2004     Star Wars  10131-110134-16964-26963-17262-16964-16963-269...   
16  2005     Star Wars  7250-17258-17257-17256-17252-17255-16966-16965...   
20  2006     Star Wars  6205-16210-16206-16207-110175-17261-26212-1621...   
24  2007     Star Wars  7660-17661-17657-110179-17665-17666-17662-1766...   
28  2008     Star Wars  10188-110186-17680-17679-17678-17676-17681-176...   
32  2009     Star Wars  20009-120007-120010-17749-17748-1comcon007-1co...   
36  2010     Star Wars  20016-12853835-18128-130050-18129-18086-18085-...   
42  2011     Star Wars  10221-15000067-120021-120019-17957-166396-1663...   
47  2012     Star Wars  9494-19675-19676-19677-19678-19679-110225-1102...   
52  2013     Star Wars  10240-1comcon032-175007-166449-15001709-150016...   
59  2014     Star Wars  75024-166515-166514-166512-1TRUXWINGTIE-150021...   
67  2015     Star Wars  5002939-15002938-175099-175096-1celeb2015-1751...   
76  2016     Star Wars  SW911617-1SW911608-111912-175118-175119-175120...   
82  2017  Super Heroes  76082-176072-171017-1830612-130607-130546-1305...   

                                               name_x  num_parts  \
0   X-wing FighterNaboo FighterTIE Fighter & Y-win...     1384.0   
2   Pit Droid™ / Star Wars Pit DroidDestroyer Droi...     2580.0   
4   Stormtrooper™TIE FighterBattle Droid CarrierIm...     2949.0   
6   Star Wars Co-Pack of 7142 and 7152Jango Fett's...     4735.0   
9   Mos Eisley Cantina, Blue boxAT-TEHailfire Droi...     6660.0   
12  TIE Fighter CollectionY-wing Attack Starfighte...     1659.0   
16  Clone Scout WalkerWookiee AttackUltimate Light...     4730.0   
20  V-wing FighterJabba's Sail BargeTIE Intercepto...     2769.0   
24  Naboo N-1 Starfighter and Vulture DroidJedi St...    11361.0   
28  Death StarGeneral GrievousThe Twilight - Limit...     6865.0   
32  AT-TE Walker - MiniRepublic Attack Cruiser - M...     3953.0   
36  Imperial Shuttle - MiniWhite Boba Fett FigureC...     6003.0   
42  Super Star DestroyerStar Wars Sith KitBounty H...     3451.0   
47  Anakin's Jedi InterceptorSebulba's Podracer & ...     6769.0   
52  Red Five X-Wing StarfighterJEK-14 Mini Stealth...     6159.0   
59  HH-87 StarhopperStar Wars Microfighters Super ...     8293.0   
67  The PhantomStormtrooper SergeantRey's Speeder™...    11410.0   
76  Palpatine's ShuttleLandspeederLEGO Star Wars: ...     6934.0   
82  ATM Heist BattleMighty Micros: Iron Man vs. Th...    13123.0   

                                           theme_name     id  is_licensed  \
0   Star Wars Episode 4/5/6Star Wars Episode 1Star...   2054           13   
2   Star WarsStar WarsStar Wars Episode 1Star Wars...   4108           26   
4   Star WarsStar Wars Episode 4/5/6Star Wars Epis...   2212           14   
6   Star Wars Episode 4/5/6Star Wars Episode 2Star...   4424           28   
9   Star Wars Episode 4/5/6Star Wars Episode 2Star...   5056           32   
12  Star Wars Episode 4/5/6Star Wars Episode 4/5/6...   3160           20   
16  Star Wars Episode 3Star Wars Episode 3Star War...   4424           28   
20  Star Wars Episode 3Star Wars Episode 4/5/6Star...   1738           11   
24  Star Wars Episode 1Star Wars Episode 3Star War...   2528           16   
28  Star Wars Episode 4/5/6Star Wars Episode 3Star...   3634           23   
32  Star Wars Episode 2Star Wars Episode 3Star War...   6162           39   
36  Star Wars Episode 4/5/6Star WarsStar Wars Clon...   4740           30   
42  Star Wars Episode 4/5/6Star WarsStar Wars Clon...   5056           32   
47  Star Wars Episode 3S

Answer: new_era = 2017

### Breakdown Number of sets by year

In [50]:
clean_df = merged[~merged['set_num'].isnull()]
clean_df['count'] = 1
sets_per_year = clean_df.groupby(['year']).sum().reset_index()[['year','count']]
#sets_per_year.head(65)

for index, rows in sets_per_year.iterrows():
    print(rows['year'], rows['count'])


1950 7
1953 4
1954 14
1955 28
1956 12
1957 21
1958 42
1959 4
1960 3
1961 17
1962 40
1963 18
1964 11
1965 10
1966 89
1967 21
1968 25
1969 69
1970 29
1971 45
1972 38
1973 68
1974 39
1975 31
1976 68
1977 92
1978 73
1979 82
1980 88
1981 79
1982 76
1983 57
1984 76
1985 139
1986 123
1987 209
1988 68
1989 114
1990 85
1991 106
1992 115
1993 111
1994 128
1995 128
1996 144
1997 194
1998 325
1999 300
2000 327
2001 339
2002 447
2003 415
2004 371
2005 330
2006 283
2007 319
2008 349
2009 403
2010 444
2011 502
2012 615
2013 593
2014 715
2015 670
2016 608
2017 438


C:\Users\manoj.c\AppData\Local\Temp\ipykernel_15828\2620521872.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_df['count'] = 1
